### PR TITLE
refactor(hashtable): adaptive string hash table

### PR DESCRIPTION
**Pull request by usamoi**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

It's the pull request about [OSPP 2022 Adaptive String Hash Table](https://github.com/datafuselabs/databend/discussions/5268).

### CODE REVIEW

**Comment by sandflee** [comment]
cool,  is there any performance stats if using `adaptive string hash table` ?

**Comment by PsiACE** [comment]
> cool, is there any performance stats if using `adaptive string hash table` ?

https://github.com/usamoi/saha/blob/main/doc/benchmark.md

**Comment by Xuanwo** [comment]
Hi, any updates on this PR? The conflict files list is longer day by day...

**Comment by PsiACE** [comment]
> Is it possible to split it into multiple so that we don't need to pending on a large PR?

This PR needs to be marged according to the requirements of the ospp assessment. But I'm not sure if @usamoi has enough energy to finish it before the dead line.

**Comment by sundy-li** [comment]
> This PR needs to be merged according to the requirements of the ospp assessment. But I'm not sure if @usamoi has enough energy to finish it before the dead line.

@usamoi told me he will submit the updates in recent days.

**Comment by sundy-li** [comment]
I just bench it with the `hits` dataset. It shows `20%~50%` performance improvement in group aggregation queries. 

<google-sheets-html-origin><style type="text/css"></style>

Query | Main | new hashtable | ratio（old/new) | comments
-- | -- | -- | -- | --
Q1 | 1.05 | 0.94 | 1.117021277 |  
Q2 | 2.9 | 1.98 | 1.464646465 | distinct int
Q3 | 1.5 | 1.3 | 1.153846154 | distinct string
Q4 | 0.05 | 0.05 | 1 |  
Q5 | 2.62 | 1.78 | 1.471910112 | group by int
Q6 | 2.79 | 1.97 | 1.416243655 | group by int
Q7 | 0.38 | 0.3 | 1.266666667 | group by string
Q8 | 0.43 | 0.34 | 1.264705882 | group by string
Q9 | 3.1 | 2.83 | 1.09540636 |  
Q10 | 3.42 | 3.42 | 1 |  
Q11 | 20.72 | 20.83 | 0.9947191551 |  
Q12 | 4.56 | 4.74 | 0.9620253165 |  
Q13 | 6.62 | 6.02 | 1.099667774 |  
Q14 | 16.98 | 15.83 | 1.072646873 |  




https://docs.google.com/spreadsheets/d/1BnNEoaA37XxhCQwzwMxp2gRK04tp8hqnwbY6m1pVT1o/edit#gid=0

File: src/common/hashtable/src/traits.rs [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+
+impl FastHash for u128 {
+    #[inline(always)]
+    fn fast_hash(&self) -> u64 {
+        cfg_if::cfg_if! {
+            if #[cfg(target_feature = "sse4.2")] {
+                use std::arch::x86_64::_mm_crc32_u64;
+                let mut high = CRC_A;
+                let mut low = CRC_B;
+                let y = [self as u64, (self >> 64) as u64];
```
Comment (sundy-li) [inline src/common/hashtable/src/traits.rs:237]:
```suggestion
                let y = [*self as u64, (self >> 64) as u64];
```

**Review by PsiACE — APPROVED** [review]

**Comment by PsiACE** [comment]
@usamoi Great job, glad to see you did it.

cc @sundy-li , Once the version is released, we can consider updating the version in ClickBench.

**Comment by Xuanwo** [comment]
@mergify update

File: src/common/hashtable/src/table0.rs [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+            }
+            if K::is_zero(&self.entries[i].key) {
+                continue;
+            }
+            let key = unsafe { self.entries[i].key.assume_init_ref() };
+            let hash = K::hash(key);
+            let index = (hash as usize) & (self.entries.len() - 1);
+            for j in (index..self.entries.len()).chain(0..index) {
+                unsafe {
+                    assume(j < self.entries.len());
```
Comment (sundy-li) [inline src/common/hashtable/src/table0.rs:228]: what's this assume used for?

**Review by sundy-li — APPROVED** [review]
Great job!

File: src/common/hashtable/src/table0.rs (hunk 2 above)
Comment (usamoi) [reply src/common/hashtable/src/table0.rs:228]: To skip bound checking in the following lines.

**Comment by BohuTANG** [comment]
Great job!

**Pull request merged**